File: C++/7562_나이트의이동.cpp
#include <iostream>
#include <queue>
#include <cstring>
using namespace std;

struct Pos{
	int x, y;
};

int main() {
	int T;
	cin >> T;
	for(int t = 0; t < T; t++){
		int I;
		cin >> I;
		
		bool **visit = new bool*[I]; // 나이트의 방문을 확인할 배열
		for(int i = 0; i < I; i++){
			visit[i] = new bool[I];
			memset(visit[i],false,sizeof(bool)*I);
		}
		
		Pos cur, goal; // 현재 위치와 최종 위치
		cin >> cur.x >> cur.y;
		cin >> goal.x >> goal.y;
		
		int dir[8][2] = {{-2,-1},{-2,1},{-1,-2},{-1,2},{2,-1},{2,1},{1,-2},{1,2}}; // 나이트의 이동 방향
		
		queue<Pos> q; // BFS로 탐색
		q.push(cur);
		visit[cur.x][cur.y] = true;
		
		int move = 0; // 이동 횟수
		bool flag = false; // goal을 만나면 true가 됨
		if(cur.x == goal.x && cur.y == goal.y){
			flag = true; // 입력받은 현재 위치와 최종 위치가 동일할 때는 이동할 필요 없음
		}
		while(!q.empty() && flag == false){
			move++;
			
			int q_len = q.size(); // move 단위로 이동할 것 따라서 queue의 크기만큼 돌면 move 1씩 증가
			for(int i = 0; i < q_len; i++){  // BFS
				cur = q.front();
				q.pop();
				
				for(int d = 0; d < 8; d++){
					int x = cur.x+dir[d][0], y = cur.y+dir[d][1];
					if(x >= 0 && x < I && y >= 0 && y < I && visit[x][y] == false){
						if(x == goal.x && y == goal.y){
							flag = true; // 최종 위치에 도착하면 true로 flag 바꾸고 break
							break;
						}
						visit[x][y] = true;
						q.push({x,y}); // 최종 위치가 아니라면 방문 표시 후 queue에 값 추가
					}
				}
			}
		}
		
		cout << move << '\n';
	}
	
	return 0;
}

File: Python/7562_나이트의이동.py
if __name__ == '__main__':
	T = int(input())
	for _ in range(T):
		I = int(input())
		visit = [[False]*I for _ in range(I)] # 나이트의 방문 표시를 위한 배열
		
		cur = list(map(int,input().split()))
		goal = list(map(int,input().split()))
		
		dir = [[-2,-1],[-2,1],[-1,-2],[-1,2],[2,-1],[2,1],[1,-2],[1,2]] # 나이트의 이동 방향
		
		flag = False # 최종 위치에 도착했는지 여부
		if cur[0] == goal[0] and cur[1] == goal[1]: # 입력받은 현재 위치와 최종 위치가 동일할 때
			flag = True
			
		time = 0 # 이동 횟수
		q = [cur+[0]] # queue는 현재 좌표 + 걸린 시간의 쌍을 가지고 있음
		while q and flag == False: # BFS 시작
			cur = q.pop(0)
			
			for d in dir: # 방향 이동
				x = cur[0]+d[0]
				y = cur[1]+d[1]
				if x >= 0 and x < I and y >= 0 and y < I and visit[x][y] == False:
					if x == goal[0] and y == goal[1]: # 이동 위치가 목적지와 동일할 때
						time = cur[2]+1 # 현재 좌표까지의 이동 횟수는 이전 이동 횟수 +1 따라서 최종 걸린 시간에 대입
						flag = True # 목적지에 도달했으므로, flag는 True로 바꿔주고 break(더 이상 탐색하지 않도록)
						break
					visit[x][y] = True
					q.append([x,y,cur[2]+1]) # 목적지가 아닐 경우에는 방문 여부 True로 변경 후 queue에 이동좌표,시간 삽입
					
		print(time)
